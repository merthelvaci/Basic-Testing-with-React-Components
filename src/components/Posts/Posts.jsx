import React, { useCallback, useReducer } from 'react';
import axios from 'axios';
import Spinner from '../Spinner';
import Post from './Post';

// reducer function which will be triggered upon each state update
export const postsReducer = (state, action) => {
	switch (action.type) {
		case 'FETCH_POSTS_START':
			return {
				...state,
				error: action.error,
				loading: action.loading,
				posts: action.posts,
			};
		case 'FETCH_POSTS_SUCCESS':
			return {
				...state,
				error: action.error,
				loading: action.loading,
				posts: action.posts,
			};
		case 'FETCH_POSTS_FAIL':
			return {
				...state,
				error: action.error,
				loading: action.loading,
				posts: action.posts,
			};
		default:
			return state;
	}
};

// initial component state to be start with
const initialPostsState = {
	posts: null,
	loading: false,
	error: null,
};

const Posts = () => {
	const [postsState, postsDispatch] = useReducer(
		postsReducer,
		initialPostsState
	);
	const { error, loading, posts } = postsState; // desctructure component state

	const fetchPostsHandler = useCallback((url) => {
		try {
			postsDispatch({
				// before even trying to fetch data, set "loading" state to true in order to indicate a spinner to user
				type: 'FETCH_POSTS_START',
				error: null,
				posts: null,
				loading: true,
			});
			const fetch = async () => {
				// define a function being responsible for making request to API to fetch required data
				const response = await axios.get(url);
				const data = response.data;
				return postsDispatch({
					// perform proper state update
					type: 'FETCH_POSTS_SUCCESS',
					error: null,
					loading: false,
					//posts: response.data,
					posts: data,
				});
			};
		} catch (e) {
			return postsDispatch({
				// in case any failure in fetching data, set "error" state to true along with proper update of other states
				type: 'FETCH_POSTS_FAIL',
				error: true,
				loading: false,
				posts: null,
			});
		}
		fetch(); // call the async function
	}, []);

	return (
		<>
			{loading && <Spinner />}
			{error && !loading && (
				<div aria-label="error_wrapper" className="error-wrapper">
					<p aria-label="error_text" className="error-text">
						An error occurred
					</p>
				</div>
			)}
			{!posts && !error && (
				<div
					aria-label="normal_content_wrapper"
					className="normal-content-wrapper"
				>
					<p
						aria-label="normal_content_text"
						className="normal-content-text"
					>
						No data fetching occurs yet
					</p>
					<button
						aria-label="fetch_data"
						onClick={() =>
							fetchPostsHandler(
								'https://jsonplaceholder.typicode.com/posts'
							)
						}
					>
						FETCH POSTS
					</button>
				</div>
			)}
			{posts &&
				posts.map((post) => (
					<Post
						key={post.body}
						body={post.body}
						id={post.id}
						title={post.title}
						userId={post.userId}
					/>
				))}
		</>
	);
};

export default Posts;
