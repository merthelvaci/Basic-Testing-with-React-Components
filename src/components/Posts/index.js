import Posts from './Posts';

export default Posts;
export { postsReducer } from './Posts';
