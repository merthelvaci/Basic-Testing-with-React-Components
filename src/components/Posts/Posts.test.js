import Posts, { postsReducer } from '.';
import axiosMock from 'axios';

let asFrag;
let byTestId;
let byText;
let _container;
beforeEach(() => {
	const { asFragment, container, getByTestId, getByText } = render(<Posts />);
	asFrag = asFragment;
	byTestId = getByTestId;
	byText = getByText;
	_container = container;
});

afterEach(cleanup);

describe('<Posts /> component tests', () => {
	it('should render <Posts /> component correctly', () => {
		expect(asFrag()).toMatchSnapshot();
	});

	describe('postsReducer function tests', () => {
		let initialState = {
			error: null,
			loading: false,
			posts: null,
		};
		it('should set error state to true upon failing data fetching', () => {
			const updatedState = postsReducer(initialState, {
				type: 'FETCH_POSTS_FAIL',
				error: true,
				loading: false,
				posts: null,
			});
			expect(updatedState.error).toBeTruthy();
			expect(updatedState.loading).toBeFalsy();
			expect(updatedState.posts).toBeFalsy();
		});

		it('should set loading state to true upon starting data fetching', () => {
			const updatedState = postsReducer(initialState, {
				type: 'FETCH_POSTS_START',
				error: null,
				posts: null,
				loading: true,
			});
			expect(updatedState.error).toBeFalsy();
			expect(updatedState.loading).toBeTruthy();
			expect(updatedState.posts).toBeFalsy();
		});

		it('should reset error state to false upon successing data fetching', () => {
			initialState = {
				error: true,
				loading: false,
				posts: null,
			};
			const updatedState = postsReducer(initialState, {
				type: 'FETCH_POSTS_SUCCESS',
				error: null,
				loading: false,
			});
			expect(updatedState.error).toBeFalsy();
			expect(updatedState.loading).toBeFalsy();
		});
	});

	it('should render no post(s) at startup', () => {
		const noPostsText = byTestId('normal_content_text');

		expect(noPostsText).toHaveTextContent('No data fetching occurs yet');
	});

	it('should have a loading spinner displayed while data fetching in progress', () => {
		const fetchBtn = byTestId('fetch_btn');
		fireEvent.click(fetchBtn);

		expect(_container.querySelector('.lds-roller')).toBeInTheDocument();
	});

	/* it('should fetch posts data upon clicking "FETCH POSTS" button', async () => {
		// Starting from Line-85 through Line-101, Enzyme is used. In order to test using React Testing Library (RTL) implementation,
		// please comment out those lines and uncomment lines starting from Line-103
		const component = mount(<Posts />);
		const fetchPostsData = axiosMock.get.mockResolvedValueOnce({
			data: {
				// fake data representation. It has these properties since the response returned from the API has this shape as well.
				body: 'Post Body 1', // in order to get more detail about fetch request, please refer to Line-111 in Posts.jsx
				id: 'Post Id 1',
				title: 'Post Title 1',
				userId: 'Post User Id 1',
			},
		});

		expect(fetchPostsData).toHaveBeenCalledTimes(0); // ensure that before firing click event, mock function will not be called
		component
			.find('[data-testid="fetch_btn"]')
			.simulate('click', fetchPostsData);

		expect(fetchPostsData).toHaveBeenCalledTimes(1); // check that mock function was triggered upon simulating a click action

		// Following code snippet is COMPLETELY WRONG! I dont know how to test async function triggered upon click event using RTL!
		const fetchButtonEl = byTestId('fetch_btn');
        fireEvent.click(fetchButtonEl);
		await axiosMock.get.mockResolvedValueOnce({
			data: {
				body: 'Post Body 1',
				id: 'Post Id 1',
				title: 'Post Title 1',
				userId: 'Post User Id 1',
			},
		});

		await waitForElement(() =>
			expect(axiosMock.get).toHaveBeenCalledTimes(1)
		);
	}); */

	it('should render a fake post data upon successful data fetching', async () => {
		axiosMock.get.mockResolvedValueOnce({
			data: {
				body: 'Post Body 1', // in order to get more detail about fetch request, please refer to Line-111 in Posts.jsx
				id: 'Post Id 1',
				title: 'Post Title 1',
				userId: 'Post User Id 1',
			},
		});

		const fetchButtonEl = byTestId('fetch_btn');
		fireEvent.click(fetchButtonEl);
		await wait(() => byTestId('post_wrapper'));

		expect(byTestId('post_title')).toHaveTextContent('Title: Post Title 1');
	});
});
