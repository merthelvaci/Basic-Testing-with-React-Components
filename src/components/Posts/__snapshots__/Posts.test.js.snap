// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts /> component tests should render <Posts /> component correctly 1`] = `
<DocumentFragment>
   
  <div
    aria-label="normal_content_wrapper"
    class="normal-content-wrapper"
  >
    <p
      aria-label="normal_content_text"
      class="normal-content-text"
    >
      No data fetching occurs yet
    </p>
    <button
      aria-label="fetch_data"
    >
      FETCH POSTS
    </button>
  </div>
</DocumentFragment>
`;
